pyqtgraph/functions.py

```diff
@@ -1752,51 +1750,121 @@ def arrayToQPath(x, y, connect='all'):
     # decide which points are connected by lines
     if eq(connect, 'all'):
-        arr[1:-1]['c'] = 1
+        polygon = arrayToQPolygonF(
+            x,
+            y
+        )
+        path.addPolygon(polygon)
+        return path
     elif eq(connect, 'pairs'):
-        arr[1:-1]['c'][::2] = 0
-        arr[1:-1]['c'][1::2] = 1  # connect every 2nd point to every 1st one
+        xs = x.tolist()
+        ys = y.tolist()
+ 
+        lineTo = path.lineTo
+        moveTo = path.moveTo
+        for elementType, x, y in zip(
+            itertools.cycle((moveTo, lineTo)),
+            xs,
+            ys
+        ):
+            elementType(x, y)
+        return path
     elif eq(connect, 'finite'):
         # Let's call a point with either x or y being nan is an invalid point.
         # A point will anyway not connect to an invalid point regardless of the
         # 'c' value of the invalid point. Therefore, we should set 'c' to 0 for
         # the next point of an invalid point.
         if isfinite is None:
             isfinite = np.isfinite(x) & np.isfinite(y)
-        arr[2:]['c'] = isfinite
+        arr[1:]['c'] = isfinite[:-1]
     elif isinstance(connect, np.ndarray):
-        arr[2:-1]['c'] = connect[:-1]
+        arr[1:]['c'] = connect[:-1]
     else:
         raise Exception('connect argument must be "all", "pairs", "finite", or array')
 
-    arr[1]['c'] = 0  # the first vertex has no previous vertex to connect
+    arr[0]['c'] = 0  # the first vertex has no previous vertex to connect
 
-    byteview.data[-20:-16] = struct.pack('>i', 0)  # cStart
-    byteview.data[-16:-12] = struct.pack('>i', 0)  # fillRule (Qt.OddEvenFill)
+    # cStart, fillRule (Qt.OddEvenFill)
+    struct.pack_into('>ii', backstore, 4+n*20, 0, 0)
 
     # create datastream object and stream into path
 
     ## Avoiding this method because QByteArray(str) leaks memory in PySide
     #buf = QtCore.QByteArray(arr.data[12:lastInd+4])  # I think one unnecessary copy happens here
 
-    path.strn = byteview.data[16:-12]  # make sure data doesn't run away
+    path.strn = backstore
     try:
         buf = QtCore.QByteArray.fromRawData(path.strn)
     except TypeError:
-        buf = QtCore.QByteArray(bytes(path.strn))
+        buf = QtCore.QByteArray(path.strn)
     except AttributeError:
         # PyQt6 raises AttributeError
-        buf = QtCore.QByteArray(path.strn, path.strn.nbytes)
+        buf = QtCore.QByteArray(path.strn, len(path.strn))
 
     ds = QtCore.QDataStream(buf)
     ds >> path
-
     return path
 
+def arrayToQPolygonF(x, y):
+    """
+    Utility function to convert two 1D-NumPy arrays representing curve data
+    (X-axis, Y-axis data) into a single open polygon (QtGui.PolygonF) object.
+    
+    Thanks to PythonQwt for making this code available
+    
+    License/copyright: MIT License © Pierre Raybaut 2020.
+
+    Parameters
+    ----------
+    x : np.array
+        x-axis coordinates for data to be plotted, must have np.float64 dtype 
```

Comment:
need to update the docs to remove the float64 requirement

Reply:
nice catch, fixed!